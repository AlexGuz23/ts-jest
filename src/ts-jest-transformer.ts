import { TsJestGlobalOptions, TsJestConfig, BabelConfig } from './types';
import TsProgram from './ts-program';
import Memoize from './utils/memoize';
import { normalizeDiagnosticTypes } from './utils/diagnostics';
import { backportJestConfig } from './utils/backports';
import jestRootDir from './utils/jest-root-dir';
import { sep, resolve } from 'path';
import parseJsonUnsafe from './utils/parse-json-unsafe';
import * as babelCfg from './utils/babel-config';
import closestPatckageJson from './utils/closest-package-json';
import sha1 from './utils/sha1';

export default class TsJestTransformer implements jest.Transformer {
  @Memoize(jestRootDir)
  sanitizedJestConfigFor<T extends jest.ProjectConfig | jest.InitialOptions>(
    jestConfig: T,
  ): T {
    const config = {
      ...(jestConfig as object),
      rootDir: jestRootDir(jestConfig),
    } as T;
    delete config.cacheDirectory;
    delete config.name;
    return config;
  }

  @Memoize(jestRootDir)
  babelJestFor(jestConfig: jest.ProjectConfig): jest.Transformer {
    // babel-jest is shipped with jest, no need to use the importer
    return require('babel-jest').createTransformer(
      this.babelConfigFor(jestConfig),
    );
  }

  @Memoize(jestRootDir)
  babelConfigFor(jestConfig: jest.ProjectConfig): BabelConfig | undefined {
    const config = this.configFor(jestConfig);
    const rootDir = jestRootDir(jestConfig);
    if (config.babelJest === false) {
      return;
    }

    let babelConfig!: BabelConfig;
    if (config.babelJest === true) {
      // lookup babelrc file
      babelConfig = babelCfg.extend({}, babelCfg.loadDefault(rootDir));
    } else if (typeof config.babelJest === 'string') {
      // path to a babelrc file
      let filePath = config.babelJest.replace('<rootDir>', `${rootDir}${sep}`);
      filePath = resolve(rootDir, filePath);
      babelConfig = parseJsonUnsafe(filePath);
    } else {
      // it's already an object with the config
      babelConfig = config.babelJest;
    }

    // ensure to return a freezed copy object
    return babelCfg.freeze(babelCfg.extend({}, babelConfig));
  }

  @Memoize(jestRootDir)
  configFor(jestConfig: jest.ProjectConfig): TsJestConfig {
    const parsedConfig = backportJestConfig(jestConfig);
    const { globals = {} } = parsedConfig as any;
    const options: TsJestGlobalOptions = { ...globals['ts-jest'] };
    return {
      inputOptions: options,
      babelJest: options.babelJest || false,
      diagnostics: normalizeDiagnosticTypes(options.diagnostics),
    };
  }

  @Memoize(jestRootDir)
  programFor(jestConfig: jest.ProjectConfig): TsProgram {
    const myConfig = this.configFor(jestConfig);
    return new TsProgram(jestRootDir(jestConfig), myConfig);
  }

  process(
    source: string,
    filePath: jest.Path,
    jestConfig: jest.ProjectConfig,
    transformOptions?: jest.TransformOptions,
  ): jest.TransformedSource | string {
    let result: string | jest.TransformedSource;

    // get the tranformer instance
    const program = this.programFor(jestConfig);
    const config = this.configFor(jestConfig);
    const instrument: boolean =
      !!transformOptions && transformOptions.instrument;

    // transpile TS code (source maps are included)
    result = program.transpileModule(filePath, source, instrument);

    // calling babel-jest transformer
    if (config.babelJest) {
      result = this.babelJestFor(jestConfig).process(
        result,
        filePath,
        jestConfig,
        transformOptions,
      );
    }

    return result;
  }

  // we can cache as for same instance the cache key won't change as soon as the path/content pair
  // doesn't change
  // TODO: find out if jest is already using this cache strategy and remove it if so
  @Memoize((data: string, path: string) => `${path}::${data}`)
  getCacheKey(
    fileContent: string,
    filePath: string,
    jestConfigStr: string,
    {
      instrument = false,
      rootDir,
    }: { instrument?: boolean; rootDir?: string } = {},
  ): string {
    const CHAR0 = '\0';
    // will be used as the hashing data source
    const hashData: string[] = [];
    const hashUpdate = (data: string) => hashData.push(data, CHAR0);

    // add file path and its content
    hashUpdate(filePath);
    hashUpdate(fileContent);

    // saniize and normalize jest config
    const jestConfig: jest.ProjectConfig = JSON.parse(jestConfigStr);
    jestConfig.rootDir = rootDir = jestRootDir({
      rootDir: rootDir || jestConfig.rootDir,
    });
    const sanitizedJestConfig: jest.ProjectConfig = this.sanitizedJestConfigFor(
      jestConfig,
    );
    // add jest config
    hashUpdate(JSON.stringify(sanitizedJestConfig));
    // add project's package.json
    const projectPkg = closestPatckageJson(rootDir, true);
    hashUpdate(projectPkg);
    // add babel config if using babel jest
    const babelConfig = this.babelConfigFor(jestConfig) || {};
    hashUpdate(JSON.stringify(babelConfig));
    // add tsconfig
    const tsConfig = this.programFor(sanitizedJestConfig).parsedConfig;
    hashUpdate(JSON.stringify(tsConfig));
    // add instrument, even if we don't use it since `canInstrument` is false
    hashUpdate(`instrument:${instrument ? 'on' : 'off'}`);

    return sha1(...hashData);
  }

  // TODO: use jest-config package to try to get current config and see if we are going to use babel jest or not
  // in which case we'd return `true` there:
  // get canInstrument() {}
}
