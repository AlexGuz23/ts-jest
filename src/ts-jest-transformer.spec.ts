import stringify from 'fast-json-stable-stringify'

import { TsJestTransformer } from './ts-jest-transformer'

describe('configFor', () => {
  it('should return the same config-set for same values', () => {
    const obj1 = { cwd: '/foo', rootDir: '/bar', globals: {} }
    const obj2 = { ...obj1 }
    const str = stringify(obj1)
    const cs1 = new TsJestTransformer().configsFor(obj1 as any)
    const cs2 = new TsJestTransformer().configsFor(obj2 as any)
    const cs3 = new TsJestTransformer().configsFor(str)
    expect(cs1.cwd).toBe('/foo')
    expect(cs1.rootDir).toBe('/bar')
    expect(cs2).toBe(cs1)
    expect(cs3).toBe(cs1)
  })
})

describe('lastTransformerId', () => {
  it('should increment for each instance', () => {
    const start = TsJestTransformer.lastTransformerId
    const id1 = new TsJestTransformer()
    const id2 = new TsJestTransformer()
    expect(id1).not.toBe(start)
    expect(id2).not.toBe(start)
    expect(id2).not.toBe(id1)
  })
})

describe('process', () => {
  let tr: TsJestTransformer
  let babel: any
  let args: [string, string, any, any]
  const config = {
    shouldStringifyContent: jest.fn(),
    get babelJestTransformer() {
      return babel
    },
    tsCompiler: { compile: jest.fn() },
    hooks: {},
  }
  const INPUT = 'export default "foo"'
  const FILE = '/foo/bar.ts'
  const JEST_CONFIG = {} as jest.ProjectConfig
  const OPTIONS = { instrument: false }
  const process = () => tr.process(...args)
  beforeEach(() => {
    tr = new TsJestTransformer()
    args = [INPUT, FILE, JEST_CONFIG, OPTIONS]
    jest
      .spyOn(tr, 'configsFor')
      .mockImplementation(() => config)
      .mockClear()
    config.shouldStringifyContent.mockImplementation(() => false).mockClear()
    babel = { process: jest.fn(s => `babel:${s}`) }
    config.tsCompiler.compile.mockImplementation(s => `ts:${s}`).mockClear()
  })

  it('should process input without babel', () => {
    babel = null
    expect(process()).toBe(`ts:${INPUT}`)
    expect(config.shouldStringifyContent.mock.calls).toMatchInlineSnapshot(`
Array [
  Array [
    "/foo/bar.ts",
  ],
]
`)
    expect(config.tsCompiler.compile.mock.calls).toMatchInlineSnapshot(`
Array [
  Array [
    "export default \\"foo\\"",
    "/foo/bar.ts",
  ],
]
`)
  })

  it('should process input with babel', () => {
    expect(process()).toBe(`babel:ts:${INPUT}`)
    expect(config.shouldStringifyContent.mock.calls).toMatchInlineSnapshot(`
Array [
  Array [
    "/foo/bar.ts",
  ],
]
`)
    expect(config.babelJestTransformer.process.mock.calls).toMatchInlineSnapshot(`
Array [
  Array [
    "ts:export default \\"foo\\"",
    "/foo/bar.ts",
    Object {},
    Object {
      "instrument": false,
    },
  ],
]
`)
    expect(config.tsCompiler.compile.mock.calls).toMatchInlineSnapshot(`
Array [
  Array [
    "export default \\"foo\\"",
    "/foo/bar.ts",
  ],
]
`)
  })

  it('should return stringified version of file', () => {
    config.shouldStringifyContent.mockImplementation(() => true)
    expect(process()).toMatchInlineSnapshot(`"ts:module.exports=\\"export default \\\\\\"foo\\\\\\"\\""`)
  })

  it('should not pass the instrument option to babel-jest', () => {
    args[3] = { instrument: true }
    expect(process()).toBe(`babel:ts:${INPUT}`)
    expect(config.babelJestTransformer.process.mock.calls).toMatchInlineSnapshot(`
Array [
  Array [
    "ts:export default \\"foo\\"",
    "/foo/bar.ts",
    Object {},
    Object {
      "instrument": false,
    },
  ],
]
`)
  })
})

describe('getCacheKey', () => {
  it('should be different for each argument value', () => {
    const tr = new TsJestTransformer()
    jest.spyOn(tr, 'configsFor').mockImplementation(jestConfigStr => ({ cacheKey: jestConfigStr }))
    const input = {
      fileContent: 'export default "foo"',
      fileName: 'foo.ts',
      jestConfigStr: '{"foo": "bar"}',
      options: { instrument: false, rootDir: '/foo' },
    }
    const keys = [
      tr.getCacheKey(input.fileContent, input.fileName, input.jestConfigStr, input.options),
      tr.getCacheKey(input.fileContent, 'bar.ts', input.jestConfigStr, input.options),
      tr.getCacheKey(input.fileContent, input.fileName, '{}', input.options),
      tr.getCacheKey(input.fileContent, input.fileName, '{}', { ...input.options, instrument: true }),
      tr.getCacheKey(input.fileContent, input.fileName, '{}', { ...input.options, rootDir: '/bar' }),
    ]
    // each key should have correct length
    for (const key of keys) {
      expect(key).toHaveLength(40)
    }
    // unique array should have same length
    expect(keys.filter((k, i, all) => all.indexOf(k) === i)).toHaveLength(keys.length)
  })
})
