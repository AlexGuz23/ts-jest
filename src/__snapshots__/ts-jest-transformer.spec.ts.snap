// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`process hoisting should hoist jest.mock calls using babel 1`] = `
"
\\"use strict\\";

jest.mock('./upper', function () {
    return function (s) {
        return s.toUpperCase();
    };
});
var __importDefault = this && this.__importDefault || function (mod) {
    return mod && mod.__esModule ? mod : { \\"default\\": mod };
};
var upper_1 = __importDefault(require(\\"./upper\\"));
var lower_1 = __importDefault(require(\\"./lower\\"));describe('hello', function () {
    test('my test', function () {
        jest.mock('./lower', function () {
            return function (s) {
                return s.toLowerCase();
            };
        });

        expect(upper_1.default('hello')).toBe('HELLO');
        expect(lower_1.default('HELLO')).toBe('hello');
    });
});
//# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbImZpbGUudHMiXSwibmFtZXMiOlsiamVzdCIsIm1vY2siLCJzIiwidG9VcHBlckNhc2UiLCJfX2ltcG9ydERlZmF1bHQiLCJtb2QiLCJfX2VzTW9kdWxlIiwidXBwZXJfMSIsInJlcXVpcmUiLCJsb3dlcl8xIiwiZGVzY3JpYmUiLCJ0ZXN0IiwidG9Mb3dlckNhc2UiLCJleHBlY3QiLCJkZWZhdWx0IiwidG9CZSJdLCJtYXBwaW5ncyI6IjtBQUNBOztBQU1BQSxLQUFLQyxJQUFMLENBQVUsU0FBVixFQUFxQixZQUFZO0FBQUUsV0FBTyxVQUFVQyxDQUFWLEVBQWE7QUFBRSxlQUFPQSxFQUFFQyxXQUFGLEVBQVA7QUFBeUIsS0FBL0M7QUFBa0QsQ0FBckY7QUFMQSxJQUFJQyxrQkFBbUIsUUFBUSxLQUFLQSxlQUFkLElBQWtDLFVBQVVDLEdBQVYsRUFBZTtBQUNuRSxXQUFRQSxPQUFPQSxJQUFJQyxVQUFaLEdBQTBCRCxHQUExQixHQUFnQyxFQUFFLFdBQVdBLEdBQWIsRUFBdkM7QUFDSCxDQUZEO0FBR0EsSUFBSUUsVUFBVUgsZ0JBQWdCSSxRQUFRLFNBQVIsQ0FBaEIsQ0FBZDtBQUNBLElBQUlDLFVBQVVMLGdCQUFnQkksUUFBUSxTQUFSLENBQWhCLENBQWQsQ0FFQUUsU0FBUyxPQUFULEVBQWtCLFlBQVk7QUFDMUJDLFNBQUssU0FBTCxFQUFnQixZQUFZO0FBR3hCWCxhQUFLQyxJQUFMLENBQVUsU0FBVixFQUFxQixZQUFZO0FBQUUsbUJBQU8sVUFBVUMsQ0FBVixFQUFhO0FBQUUsdUJBQU9BLEVBQUVVLFdBQUYsRUFBUDtBQUF5QixhQUEvQztBQUFrRCxTQUFyRjs7QUFGQUMsZUFBT04sUUFBUU8sT0FBUixDQUFnQixPQUFoQixDQUFQLEVBQWlDQyxJQUFqQyxDQUFzQyxPQUF0QztBQUNBRixlQUFPSixRQUFRSyxPQUFSLENBQWdCLE9BQWhCLENBQVAsRUFBaUNDLElBQWpDLENBQXNDLE9BQXRDO0FBRUgsS0FKRDtBQUtILENBTkQiLCJmaWxlIjoiZmlsZS50cyIsInNvdXJjZXNDb250ZW50IjpbIlxuXCJ1c2Ugc3RyaWN0XCI7XG52YXIgX19pbXBvcnREZWZhdWx0ID0gKHRoaXMgJiYgdGhpcy5fX2ltcG9ydERlZmF1bHQpIHx8IGZ1bmN0aW9uIChtb2QpIHtcbiAgICByZXR1cm4gKG1vZCAmJiBtb2QuX19lc01vZHVsZSkgPyBtb2QgOiB7IFwiZGVmYXVsdFwiOiBtb2QgfTtcbn07XG52YXIgdXBwZXJfMSA9IF9faW1wb3J0RGVmYXVsdChyZXF1aXJlKFwiLi91cHBlclwiKSk7XG52YXIgbG93ZXJfMSA9IF9faW1wb3J0RGVmYXVsdChyZXF1aXJlKFwiLi9sb3dlclwiKSk7XG5qZXN0Lm1vY2soJy4vdXBwZXInLCBmdW5jdGlvbiAoKSB7IHJldHVybiBmdW5jdGlvbiAocykgeyByZXR1cm4gcy50b1VwcGVyQ2FzZSgpOyB9OyB9KTtcbmRlc2NyaWJlKCdoZWxsbycsIGZ1bmN0aW9uICgpIHtcbiAgICB0ZXN0KCdteSB0ZXN0JywgZnVuY3Rpb24gKCkge1xuICAgICAgICBleHBlY3QodXBwZXJfMS5kZWZhdWx0KCdoZWxsbycpKS50b0JlKCdIRUxMTycpO1xuICAgICAgICBleHBlY3QobG93ZXJfMS5kZWZhdWx0KCdIRUxMTycpKS50b0JlKCdoZWxsbycpO1xuICAgICAgICBqZXN0Lm1vY2soJy4vbG93ZXInLCBmdW5jdGlvbiAoKSB7IHJldHVybiBmdW5jdGlvbiAocykgeyByZXR1cm4gcy50b0xvd2VyQ2FzZSgpOyB9OyB9KTtcbiAgICB9KTtcbn0pO1xuIl19"
`;

exports[`process stringifyContentPathRegex should create a module with stringified content as export 1`] = `"module.exports=\\"\\\\n<div>\\\\n  <span>some text with \`backtilt\`</span>\\\\n</div>\\\\n\\""`;
